## Pythong Fundamentals - Exercise 4.5

## 01. Importing Libraries and Create Paths

In [1]:
#Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Path Shortcut
path = r'/Users/puneet/Desktop/Instacart Basket Analysis 08-2025'

In [3]:
#Connect products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data','Original Data','products.csv'),index_col = False)

In [4]:
#Connect orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data','orders_wrangled.csv'),index_col = False)

## 02. Data Consistency

In [5]:
#Data check
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_prior_order  
count            3.421083e+06       3.421083e+06            3.214874e+06  
mean             2.776219e+00       1.345202e+01            1.111484e+01  
std              2.046829e+00       4.226088e+00            9.206737e+00  
min              0.000000e+00       0.000000e+00            0.000000e+00  
25%              1.000000e+00       1.000000e+01            4.000000e+00  
50%              3.000000e+00       1.300000e+01            7.000000e+00  
75%              5.000000e+00       1.600000e+01            1.500000e+01  
max              6.000000e+00       2.300000e+01            3.000000e+01

## -

## Mixed-Type Data
    - Column that could contain strings and numeric values

In [6]:
#Creating dataframe
df_test = pd.DataFrame()

In [7]:
#Create a mixed type column
df_test['mix'] = ['a','b',1,True]
#This column has 2 string, 1 numeric, and one boolean value (mixed)

In [8]:
#Check new dataframe
df_test.head()

mix
0     a
1     b
2     1
3  True

--

In [9]:
#Advanced Python code, this checks for mixed-type columns

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


/var/folders/_t/l5t5ly6n7sng4v9pk5llghjh0000gn/T/ipykernel_28983/4168849639.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


- Code above is using a "for-loop". "for" stands for "for these elements, do this"
- "weird" variable checks if data is consistent. If True, there is inconsistency, if False, then one data type
- Boolean values also represetned by numbers. 0 = False, 1 = True

In [10]:
#This will change 'Mix' column to a string instead of a mixed-type column
df_test['mix'] = df_test['mix'].astype('str')

## -

## Missing Values
    - First find, then fix

In [11]:
#Finding missing/null values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

- isnull() function finds a vlue of true or false, if there is a null value
- .sum() function adds up the number of "True" values

Next step is to find these 16 missing values by creating a new dataframe

In [12]:
#New df_nan dataframe which shows only null value rows
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
#Check df_nan
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Address Missing Values
1. Create a new variable that acts like a flag based on missing valuse
2. Impute the value with the means or median of the column(if numeric)
4. Remove or filter out

In [14]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Filling Column with mean/median values function:
- df['column with missings'].fillna(mean value, inplace = True)
- df['column with missings'].fillna(median value, inplace = True)

Cannot use mean values here because 'product_name' is a string

In [15]:
#Check size of current table, before removing values and comparing
df_prods.shape

(49693, 5)

In [16]:
#New clean dataframe with 16 null rows taken out
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
#Check new clean dataframe for row count
df_prods_clean.shape

(49677, 5)

- Another drop function
    - df_prods.dropna(inplace = True)
    - df_prods.dropna(subset = [‘product_name’], inplace = True)

This option lets you overwrite, instead of creating a new dataframe

## -

## Duplicates
- Find duplicates first using below function

In [18]:
#Create new 'dups' dataframe to locate duplicates first
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
#Check new dups dataframe
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

Next step is addressing duplicates
- Dropping duplicates in pandas has a function
- df.drop_duplicates() is the function

In [20]:
#check row count of dataframe before droping duplicates
df_prods_clean.shape

(49677, 5)

In [21]:
#Create new dataframe that doesn't include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
#Check row count of de-deuped dataframe
df_prods_clean_no_dups.shape

(49672, 5)

5 fewer rows, which means new dataframe has the de-duped data

## -

## Tidying Up and Exporting Changes

In [23]:
#Export new clean products dataframe as .csv in files
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'products_checked.csv'))

## ----

## Exercise 4.5

2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [24]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_prior_order  
count            3.421083e+06       3.421083e+06            3.214874e+06  
mean             2.776219e+00       1.345202e+01            1.111484e+01  
std              2.046829e+00       4.226088e+00            9.206737e+00  
min              0.000000e+00       0.000000e+00            0.000000e+00  
25%              1.000000e+00       1.000000e+01            4.000000e+00  
50%              3.000000e+00       1.300000e+01            7.000000e+00  
75%              5.000000e+00       1.600000e+01            1.500000e+01  
max              6.000000e+00       2.300000e+01            3.000000e+01

- The Max order number is 100 (maybe a cap on orders?), while the minimum is correctly 1
- The Days of the week don't look suspicious, days are 0-6, either Monday-Sunday or Sunday-Saturday as the Max/Min
- Order hour of the day is similar to day of the week. Hours go 0-23, for a 24 hours period.
- Days since prior visit has slightly fewer rows count, and may have a cap of 30 days (Max). The minimum is 0, meaning some place orders twice in the same day

3. Check for mixed-type data in your df_ords dataframe

In [26]:
#Using code from exercise above

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

/var/folders/_t/l5t5ly6n7sng4v9pk5llghjh0000gn/T/ipykernel_28983/3325542213.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/_t/l5t5ly6n7sng4v9pk5llghjh0000gn/T/ipykernel_28983/3325542213.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/_t/l5t5ly6n7sng4v9pk5llghjh0000gn/T/ipykernel_28983/3325542213.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/_t/l5t5ly6n7sng4v9pk5llghjh0000gn/T/ipykernel_28983/3325542213.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].ilo

Recieved this message, but don't think there mixed-type data columns

In [28]:
#Checking columns another way
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_the_week  int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


We see only integer column types. No mixed data

5. Run a check for missing values in your df_ords dataframe.
6. Address the missing values

In [30]:
#Find missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

We see the 'days since prior order' has 206,209 missing values. We can investigate further by checking with a new dataframe

In [31]:
#New dataframe with only missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [34]:
#Show results of new dataframe (35 rows)
df_ords_nan.head(35)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0             0   2539329        1             1                       2   
11           11   2168274        2             1                       2   
26           26   1374495        3             1                       1   
39           39   3343014        4             1                       6   
45           45   2717275        5             1                       3   
50           50   2086598        6             1                       5   
54           54   2565571        7             1                       3   
75           75    600894        8             1                       6   
79           79    280530        9             1                       1   
83           83   1224907       10             1                       2   
89           89   2411567       11             1                       0   
97           97   1483446       12             1                       1   
103         103   2618231       13             1                       6   
116         116   2906263       14             1                       2   
130         130   2514324       15             1                       5   
153         153   1273676       16             1                       2   
160         160   1737705       17             1                       2   
201         201   2780889       18             1                       6   
208         208   1321974       19             1                       6   
218         218    947722       20             1                       1   
223         223   2625565       21             1                       1   
257         257   2300537       22             1                       3   
273         273   1792521       23             1                       5   
278         278    611214       24             1                       3   
297         297   2610576       25             1                       1   
301         301   3229714       26             1                       2   
314         314    361493       27             1                       3   
396         396   1138968       28             1                       0   
421         421   2064090       29             1                       5   
440         440    375303       30             1                       1   
449         449   2150649       31             1                       5   
470         470    134879       32             1                       3   
476         476   2906103       33             1                       3   
480         480   2383661       34             1                       4   
486         486   1269089       35             1                       5   

     order_hour_of_day  days_since_prior_order  
0                    8                     NaN  
11                  11                     NaN  
26                  14                     NaN  
39                  11                     NaN  
45                  12                     NaN  
50                  18                     NaN  
54                   9                     NaN  
75                   0                     NaN  
79                  17                     NaN  
83                  14                     NaN  
89                  11                     NaN  
97                  12                     NaN  
103                 12                     NaN  
116                 18                     NaN  
130                 13                     NaN  
153                  8                     NaN  
160                 13                     NaN  
201                 18                     NaN  
208                 13                     NaN  
218                 10                     NaN  
223                 10                     NaN  
257                  7                     NaN  
273                 14                     NaN  
278                 20                     NaN  
297                 14                     

These results show us that the first order for every customer displays the 'NaN' value due to it being the first order. Therefore there cannot be a 'prior order', so that data is valid and should be kept.

7. Run a check for duplicate values in your df_ords data.
8. Address the duplicates using an appropriate method.

In [35]:
#Check for duplicates by creating new dataframe with only duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [36]:
#Show new duplicate dataframe
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates were found.

In [38]:
df_ords.shape

(3421083, 7)

In [39]:
df_ords_dups.shape

(0, 7)

If we needed to address duplicates, we would use the shape function above and make sure the correct number of rows would be taken out. We would create a new de-duped dataframe using the .drop_duplicates() function as shown in the exercise earlier

9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [40]:
#Export new clean products dataframe as .csv in files
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'products_checked.csv'))

In [42]:
#Export new clean orders dataframe as .csv in files
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data','orders_checked.csv'))